app/deployment/app.go

```diff
@@ -245,6 +373,252 @@ func (a *app) doDeliverTx(ctx apptypes.Context, tx *types.TxCreateDeployment) tm
 	}
 
 	return tmtypes.ResponseDeliverTx{
-		Tags: apptypes.NewTags(a.Name(), apptypes.TxTypeDeployment),
+		Tags: apptypes.NewTags(a.Name(), apptypes.TxTypeCreateDeployment),
+	}
+}
+
+func (a *app) doDeliverCloseTx(ctx apptypes.Context, tx *types.TxCloseDeployment) tmtypes.ResponseDeliverTx {
+
+	cresp := a.doCheckCloseTx(ctx, tx)
+	if !cresp.IsOK() {
+		return tmtypes.ResponseDeliverTx{
+			Code: cresp.Code,
+			Log:  cresp.Log,
+		}
+	}
+
+	deployment, err := a.State().Deployment().Get(tx.Deployment)
+	if err != nil {
+		return tmtypes.ResponseDeliverTx{
+			Code: code.INVALID_TRANSACTION,
+			Log:  err.Error(),
+		}
+	}
+	if deployment == nil {
+		return tmtypes.ResponseDeliverTx{
+			Code: code.INVALID_TRANSACTION,
+			Log:  "Deployment not found",
+		}
+	}
+
+	deployment.State = types.Deployment_CLOSING
+
+	for i, group := range deployment.Groups {
+		// begin for each group
+		if group.State != types.DeploymentGroup_CLOSING {
+			group.State = types.DeploymentGroup_CLOSING
+			deployment.Groups[i] = group
+		}
+
+		orders, err := a.State().Order().ForGroup(&group)
+		if err != nil {
+			return tmtypes.ResponseDeliverTx{
+				Code: code.INVALID_TRANSACTION,
+				Log:  err.Error(),
+			}
+		}
+
+		for _, order := range orders {
+			// begin for each otder
+			if order.State != types.Order_CLOSING {
+				order.State = types.Order_CLOSING
+			}
+
+			fulfillments, err := a.State().Fulfillment().ForOrder(order)
+			if err != nil {
+				return tmtypes.ResponseDeliverTx{
+					Code: code.INVALID_TRANSACTION,
+					Log:  err.Error(),
+				}
+			}
+
+			for _, fulfillment := range fulfillments {
+				// begin for each fulfillment
+				if fulfillment.State != types.Fulfillment_CLOSING {
+					fulfillment.State = types.Fulfillment_CLOSING
+				}
+
+				lease, err := a.State().Lease().Get(deployment.Address, group.Seq, order.Order, fulfillment.Provider)
+				if err != nil {
+					return tmtypes.ResponseDeliverTx{
+						Code: code.INVALID_TRANSACTION,
+						Log:  err.Error(),
+					}
+				}
+				if lease != nil && lease.State != types.Lease_CLOSING {
+					return tmtypes.ResponseDeliverTx{
+						Code: code.INVALID_TRANSACTION,
+						Log:  "Lease not closed",
+					}
+				}
```

Comment:
these early returns are inviting inconsistent state.   we need to get that transactional db support soon.